prql/src/semantic/analyzer.rs

```diff
@@ -0,0 +1,276 @@
+    // save functions declarations
+    fn fold_nodes(&mut self, items: Vec<Node>) -> Result<Vec<Node>> {
+        // We cut out function def, so we need to run it
+        // here rather than in `fold_func_def`.
+        let mut r = Vec::with_capacity(items.len());
+
+        for item in items {
+            match item {
+                Node {
+                    item: Item::FuncDef(func_def),
+                    ..
+                } => {
+                    let func_def = fold_func_def(self, func_def)?;
+
+                    self.declare_func(func_def);
+                }
+                _ => r.push(self.fold_node(item)?),
+            }
+        }
+        Ok(r)
+    }
+
+    fn fold_node(&mut self, mut node: Node) -> Result<Node> {
+        node.item = match node.item {
+            Item::FuncCall(func_call) => {
+                node.declared_at = self.last.functions.get(&func_call.name).cloned();
+
+                Item::FuncCall(self.fold_func_call(func_call)?)
+            }
+
+            Item::Ident(ident) => {
+                node.declared_at = self.lookup_variable(&ident);
+
+                Item::Ident(self.fold_ident(ident)?)
+            }
+
+            // Item::InlinePipeline(p) => self.inline_pipeline(p)?,
+
+            // Item::Ident(ident) => {
+            //     if let Some(def) = self.functions_no_args.get(ident.as_str()) {
+            //         def.body.item.clone()
+            //     } else {
+            //         Item::Ident(ident)
+            //     }
+            // }
+            item => fold_item(self, item)?,
+        };
+        Ok(node)
+    }
+
+    fn fold_pipeline(&mut self, pipeline: Vec<Transformation>) -> Result<Vec<Transformation>> {
+        pipeline
+            .into_iter()
+            .filter_map(|t| {
+                if let Transformation::Select(_) = t {
+                    self.last.clear_table();
+                }
+                let trans_name = t.name();
+
+                match t {
+                    Transformation::Select(nodes) | Transformation::Derive(nodes) => {
+                        for (position, node) in nodes.into_iter().enumerate() {
+                            let node = self.fold_node(node).unwrap();
+
+                            self.declare_table_column(position, &node);
+                        }
+                        println!("{trans_name}: {:?}", self.get_table_columns());
+                        None
+                    }
+                    t => Some(fold_transformation(self, t)),
+                }
+            })
+            .collect()
+    }
+}
+
+/// Scope within which we can reference variables, functions and tables
+/// Provides fast lookups for different names.
+#[derive(Debug, Default)]
+pub struct Scope {
+    // current table (result of last pipeline)
+    table: Vec<usize>,
+
+    // For each namespace (table), a list of its variables (columns)
+    // "" is default namespace
+    // "%" is namespace of functions without arguments
```

Comment:
```suggestion
    // "%" is namespace of functions without parameters
```
(arguments are in the call, params in the defn)